Provide an abstract output guardrails for json data extraction

mariofusco opened the pull request:
This is a first implementation of @lordofthejars 's idea discussed here https://github.com/quarkiverse/quarkus-langchain4j/discussions/1022

I believe this could be basically used for each `AiService` trying to extract a POJO from the user prompt like

```
@RegisterAiService
@Singleton
public interface CustomerExtractor {

    @UserMessage("Extract information about a customer from this text '{text}'. The response must contain only the JSON with customer's data and without any other sentence.")
    @OutputGuardrails(CustomerExtractionOutputGuardrail.class)
    Customer extractData(String text);
}
```

implementing the output guardrail simply extending the provided `AbstractJsonExtractorOutputGuardrail` as it follows

```
@ApplicationScoped
public class CustomerExtractionOutputGuardrail extends AbstractJsonExtractorOutputGuardrail {

    @Override
    protected Class<?> getOutputClass() {
        return Customer.class;
    }
}
```

/cc @geoand @cescoffier

lordofthejars commented:
Oh, wow, that is really nice. I have a list of guardrails that we could provide. I basically followed the LLM-guard project to check what guardrails they are offering and which of them we could implement.

mariofusco commented:
> Oh, wow, that is really nice. 

This is only the more straightforward things that it comes me to mind. We could add more features and implementations incrementally with other commits.

> I have a list of guardrails that we could provide. I basically followed the LLM-guard project to check what guardrails they are offering and which of them we could implement.

If you already have a list of guardrails that we could implement and provide out-of-the-box please feel free to share it, I'd be glad to collaborate on this.

lordofthejars commented:
Mario are you up next week for a quick meeting and I explain my idea?

geoand commented:
@gsmet when you have some time, I would love to chat with you about setting up a bot for this repo :)
A good starting point would be providing the awesome CI summary we have in the main Quarkus repo

gsmet commented:
I think we could install the Quarkus one here and only enable some of its features (that's something we can do).

geoand commented:
That would be awesome, thanks!

cescoffier commented on core/runtime/src/main/java/io/quarkiverse/langchain4j/guardrails/AbstractJsonExtractorOutputGuardrail.java:
```diff
@@ -0,0 +1,38 @@
+
+public abstract class AbstractJsonExtractorOutputGuardrail implements OutputGuardrail {
+
+    private static final ObjectMapper MAPPER = new ObjectMapper();
```
I'm not sure we want to do that.
Mapper are huge, you probably want to reuse an existing one, maybe even the Quarkus one.

lordofthejars replied:
`@Inject ObjectMapper MAPPER`

cescoffier replied:
Yes, that would be better (and it will also handle the modules the user has registered)

mariofusco replied:
Sorry, this is a leftover of my original implementation, this mapper is totally useless now and I will remove it.

cescoffier commented on core/runtime/src/main/java/io/quarkiverse/langchain4j/guardrails/GuardrailsUtils.java:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    private static class MapperHolder {
+        private static final ObjectMapper MAPPER = new ObjectMapper();
```
Another mapper? As I sid, we should avoid instantiating these.

mariofusco replied:
I will try to replace it with the one retrieved through Quarkus injection.

cescoffier replied:
You can use programmatic lookup and only create one if none are available.

cescoffier commented on core/runtime/src/main/java/io/quarkiverse/langchain4j/guardrails/GuardrailsUtils.java:
```diff
@@ -0,0 +1,31 @@
+
+    static boolean validateJson(String json, Class<?> expectedOutputClass) {
+        try {
+            MapperHolder.MAPPER.readValue(json, expectedOutputClass);
```
You are immediately going into Databind. There are various issues:

1. I believe you need to verify it's valid JSON before mapping (see the example from the doc)
2. You pass a Class instead of a Type/TypeReference, so you won't handle generics and collections 
3. Mapping can be super expensive, and Quarkus is going to do it anyway latter anyway

mariofusco replied:
> You are immediately going into Databind. There are various issues:
> 
>     1. I believe you need to verify it's valid JSON before mapping (see the example from the doc)

Mapping is a stronger constraint and then a super set of a valid JSON. I was thinking that in this way I'm doing both things in one shoot, but I'm no longer sure if this is a good idea. However if the JSON is valid (and I guess that in most cases it will be) I will have to also perform a mapping anyway.

>     2. You pass a Class instead of a Type/TypeReference, so you won't handle generics and collections

Good point, I will fix this.
 
>     3. Mapping can be super expensive, and Quarkus is going to do it anyway latter anyway

That's another good point, but with more complex solution. Ideally if the mapping is successful we may want to pass downstream the deserialized object instead of having to redo the mapping once again, but at the moment output guardrails don't allow to do this. I will check if I can change them a bit to make this possible.

cescoffier replied:
Totally agree with your last point - if the mapping is successful we should pass that object as response.

mariofusco replied:
> yes that would be great specially in terms of performance

I already implemented this with my latest commit, give a look.

mariofusco commented on core/runtime/src/main/java/io/quarkiverse/langchain4j/guardrails/GuardrailsUtils.java:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    static String trimNonJson(String llmResponse) {
+        int jsonStart = llmResponse.indexOf("{");
```
Good point, I will fix it.

mariofusco commented:
@cescoffier I should have addressed all your remarks. For now I'm keeping the 2 commits separated only to make it easier for you to review my latest change. In particular now an `OutputGuardrail` can not only rewrite the textual LLM response, but also provide an explicit object result for the whole AiService invocation. In this way the object eventually deserialized from the json response can be added to the result and returned as it is instead of requiring a second deserialization.

cescoffier commented on core/runtime/src/main/java/io/quarkiverse/langchain4j/guardrails/AbstractJsonExtractorOutputGuardrail.java:
```diff
@@ -0,0 +1,60 @@
+                        + "the specified format");
+    }
+
+    private Object deserialize(String llmResponse) {
```
Should we make this one protected?

mariofusco replied:
Done.

cescoffier commented on core/runtime/src/main/java/io/quarkiverse/langchain4j/guardrails/AbstractJsonExtractorOutputGuardrail.java:
```diff
@@ -0,0 +1,60 @@
+        return null;
+    }
+
+    protected TypeReference<?> getOutputType() {
```
Wondering if we can use `Type` to merge both the output class and output type?

mariofusco replied:
I tried to, but `TypeReference` doesn't extends `Type`, but wraps it so I didn't find a reasonable way to do this.

cescoffier commented on core/runtime/src/main/java/io/quarkiverse/langchain4j/guardrails/GuardrailsUtils.java:
```diff
@@ -0,0 +1,47 @@
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+@ApplicationScoped
+class GuardrailsUtils {
```
Should it be named `JsonGuardrailsUtils`?

mariofusco replied:
Done.

cescoffier reviewed (changes requested):
LGTM, I've just a few minor comments.

mariofusco commented:
I implemented all the requested changes and squashed all commits.

cescoffier commented:
We should extend a bit the guardrail documentation to remove the json example and mention this one (with an example)

mariofusco commented:
> We should extend a bit the guardrail documentation to remove the json example and mention this one (with an example)

Agreed. I will rework docs as soon as this will be merged.
